logsmanagement/parser.c

```diff
@@ -0,0 +1,1618 @@
+    "$remote_addr - - [$time_local] \"$request\" $status $body_bytes_sent $request_length $request_time",                                                // csvCustom1
+    "$remote_addr - - [$time_local] \"$request\" $status $body_bytes_sent",                                                                              // csvCommon
+    NULL}
+;
+
+typedef struct log_line_parsed{
+    char vhost[VHOST_MAX_LEN];
+    int  port;
+    char req_scheme[REQ_SCHEME_MAX_LEN];
+    char req_client[REQ_CLIENT_MAX_LEN];
+    char req_method[REQ_METHOD_MAX_LEN];
+    char req_URL[REQ_URL_MAX_LEN];
+    char req_proto[REQ_PROTO_MAX_LEN];
+    int req_size;
+    int req_proc_time;
+    int resp_code;
+    int resp_size;
+    int ups_resp_time;
+    char ssl_proto[SSL_PROTO_MAX_LEN];
+    char ssl_cipher[SSL_CIPHER_SUITE_MAX_LEN];
+    int64_t timestamp;
+    int parsing_errors;
+} Log_line_parsed_t;
+
+static inline int count_fields(const char *line, const char delimiter){
+    const char *ptr;
+    int cnt, fQuote;
+
+    for (cnt = 1, fQuote = 0, ptr = line; *ptr != '\n' && *ptr != '\r' && *ptr != '\0'; ptr++ ){
+        if (fQuote) {
+            if (*ptr == '\"') {
+                if ( ptr[1] == '\"' ) {
+                    ptr++;
+                    continue;
+                }
+                fQuote = 0;
+            }
+            continue;
+        }
+
+        if(*ptr == '\"'){
+            fQuote = 1;
+            continue;
+        }
+        if(*ptr == delimiter){
+            cnt++;
+            while(*(ptr+1) == delimiter){ ptr++;};
+            continue;
+        }
+    }
+
+    if (fQuote) {
+        return -1;
+    }
+
+    return cnt;
+}
+
+/**
+ * @brief Parse a delimited string into an array of strings.
+ * @details Given a string containing no linebreaks, or containing line breaks
+ * which are escaped by "double quotes", extract a NULL-terminated
+ * array of strings, one for every delimiter-separated value in the row.
+ * @param[in] line The input string to be parsed.
+ * @param[in] delimiter The delimiter to be used to split the string.
+ * @param[in] num_fields The expected number of fields in \p line. If a negative
+ * number is provided, they will be counted.
+ * @return A NULL-terminated array of strings with the delimited values in \p line,
+ * or NULL in any other case.
+ * @todo This function is not benchmarked or optimised.
+ */
+static inline char **parse_csv( const char *line, const char delimiter, int num_fields) {
+    char **buf, **bptr, *tmp, *tptr;
+    const char *ptr;
+    int fQuote, fEnd;
+
+    if(num_fields < 0){
+        num_fields = count_fields(line, delimiter);
+
+        if ( num_fields == -1 ) {
+            return NULL;
+        }
+    }
+
+    buf = mallocz( sizeof(char*) * (num_fields+1) );
+
+    if ( !buf ) {
+        return NULL;
+    }
+
+    tmp = mallocz( strlen(line) + 1 );
```

Comment:
this test is not necessary, because `mallocz` on fail will close netdata.

Reply:
Done.